AZURE/create.yml

```diff
@@ -0,0 +1,271 @@
+        mode: 0600
+      when: keypair.changed | bool
+
+    # During further VMs creation, we have to pass ssh_public_keys param for Linux VMs
+    - name: Set ssh_public_keys list of dicts for Linux VMs
+      set_fact:
+        ssh_public_keys:
+          - path: "/home/{{ vm_admin_username }}/.ssh/authorized_keys"
+            key_data: "{{ keypair.public_key }}"
+
+    # Get unique idempotent name for VM, in Azure VM name == ID
+    # for Windows VM name cannot be longer than 15 characters
+    # VM name sample "tst-c03a89e9-vm"
+    - name: Set unique VM name dicts
+      set_fact:
+        unique_vm_name: >-
+          {{ unique_vm_name | default({})
+            | combine({ item.name: 'tst-' + ((item.name + '-' + ci_branch_name) | to_uuid)[:8] }) }}
+      loop: "{{ molecule_yml.platforms }}"
+
+    # NICs must be created before VMs because of limitations applied by EPAM
+    # During VM creation, Azure tries to create NIC with new network security group, but we can't create NSGs, don't have such permissions
+    # Therefore, we have to create NIC first, attach default EPAM NSG to NIC, then add an existing NIC to new VM
+    - name: Create Azure network interface
+      azure_rm_networkinterface:
+        name: "{{ unique_vm_name[item.name] }}-nic"
+        resource_group: "{{ item.az_resource_group_name }}"
+        virtual_network: "{{ item.az_virtual_network_name }}"
+        subnet_name: "{{ item.az_subnet_name }}"
+        location: "{{ item.az_location }}"
+        create_with_security_group: false
+        security_group: "{{ item.az_network_security_group }}"
+        ip_configurations:
+          - name: ipconfig1
+            primary: True
+            public_ip_address_name: "{{ unique_vm_name[item.name] }}-ip"
+        tags:
+          Instance: "{{ item.name }}"
+          CI_job: "{{ item.name }}-{{ ci_branch_name }}"
+          Name: "{{ unique_vm_name[item.name] }}"
+          Created_time: "{{ lookup('pipe','date +%Y-%m-%d') }}"
+          Managed_by: "Molecule"
+      register: nics
+      loop: "{{ molecule_yml.platforms }}"
+      async: 7200
+      poll: 0
+
+    - name: Wait for NICs creation to complete
+      async_status:
+        jid: "{{ item.ansible_job_id }}"
+      loop: "{{ nics.results }}"
+      register: nic_create
+      until: nic_create.finished
+      delay: 10
+      retries: 300
+
+    - name: Add tags to public IP address
+      azure_rm_publicipaddress:
+        name: "{{ item.state.ip_configurations[0].public_ip_address.name }}"
+        resource_group: "{{ item.item.item.az_resource_group_name }}"
+        tags:
+          Instance: "{{ item.item.item.name }}"
+          CI_job: "{{ item.item.item.name }}-{{ ci_branch_name }}"
+          Name: "{{ unique_vm_name[item.item.item.name] }}"
+          Created_time: "{{ lookup('pipe','date +%Y-%m-%d') }}"
+          Managed_by: "Molecule"
+      loop: "{{ nic_create.results }}"
+      register: ip_tags
+      async: 7200
```

Comment:
Is this task running so long we need async?

Reply:
Yes it is. Unfortunately, every operation with resurs takes time in Azure, even adding tags to the resource :(